### ISSUE TITLE
versioning of HDF5 files (H5CUDS)

### ISSUE THREAD

**Issue by nathanfranklin** [simphony/simphony-common#156]
The file format should be extended with some version information and then to use that information for better error messages etc..  

see simphony/simphony-openfoam#26

**Comment by itziakos** [comment]
Yes versioned hdf5 layout is something very important, I suggest that this issue get a high priority.

Please note that version should have two levels. A version for the H5CUDS layout, and a separate version for each CUDS container type. The best solution is to have this added as an cuds_version attribute in the top level group of the H5CUDS group and the CUDS container's group.

**Comment by itziakos** [comment]
**Design proposal**

*Versioning*

In CUDS HDF5 layout version is provided in two levels:

1. Top level version of the main group (i.e. root) where the container types are grouped. This is the expected layout of the file as defined in H5CUDS.
2. The CUDS container hdf5 layout version separate for each CUDS container. 

The version should be stored in the top level group node as a node attribute names `cuds_version` and the value should be a single integer.

*Behaviour*

When a H5 object is asked to read from from a file or node (for CUDS containers) it will check the version number of raise if that particular version layout is not supported. 

*Backwords compativility*

To simplify the HDF5 cuds implementation and testing:
- Only a single individual version (the latest) should be supported for live (i.e. online) read/write of H5 top level objects from a file. 
- Backwards compatibility will be provided only at offline operation by separate helper conversion script that will know how to convert between versions that arrive as part of a stable release.
- Version numbers are incremented only in once in a *simphony-common* release cycle, thus intermediate changes of the HDF5 layout during the development process of a new release will only change HDF5 layout versions only once. 
- Optionally, The version number of the H5CUDS file is incremented when the main layout changes or the layout of the cuds containers change. This will simplify the implementation of the conversion script.

**Comment by nathanfranklin** [comment]
Sounds good.

 > Optionally, The version number of the H5CUDS file is incremented when the main layout changes or the layout of the cuds containers change. This will simplify the implementation of the conversion scrip

I prefer the above approach to changing the version numbers.

**Comment by nathanfranklin** [comment]
> In CUDS HDF5 layout version is provided in two levels:
>   Top level version of the main group (i.e. root) where the container types are grouped. This is the expected layout of the file as defined in H5CUDS.
>    The CUDS container hdf5 layout version separate for each CUDS container.

As you mentioned, this gives us an integer attribute called `cuds_version` for
 * main group (root)
 * particles
 * mesh
 * lattice

We probably need an additional  "cuds_version" for the two storage layouts for CUBA values (i.e. `IndexedDataContainer` and `DataContainerTable`), right?  

Changes to the list of CUBA keywords (`cuba.yml`) will affect both `IndexedDataContainer` and `DataContainerTable`.  Such changes to the CUBA list should occur regularly with future releases. 

Should we address that in this issue or open up another issue?

We could for example add a version to cuba.yml and the `cuds_version` for `IndexedDataContainer` and `DataContainerTable` could follow that number.

**Comment by itziakos** [comment]
Hi Nathan,

Please find my comments inplace,

2015-07-06 16:31 GMT+01:00 Nathan Franklin <notifications@github.com>:

> In CUDS HDF5 layout version is provided in two levels:
> Top level version of the main group (i.e. root) where the container types
> are grouped. This is the expected layout of the file as defined in H5CUDS.
> The CUDS container hdf5 layout version separate for each CUDS container.
>
> As you mentioned, this gives us an integer attribute called cuds_version
> for
>
>    - main group (root)
>    - particles
>
>
>    - mesh
>    - lattice
>
> We probably need an additional "cuds_version" for the two storage layouts
> for CUBA values (i.e. IndexedDataContainer and DataContainerTable), right?
>
I do not think that we need to version at that level. If these
corresponding layout change then we update the effected container's
versions (since it will probably effect all of them)

>  Changes to the list of CUBA keywords (cuba.yml) will affect both
> IndexedDataContainer and DataContainerTable. Such changes to the CUBA
> list should occur regularly with future releases.
>
IndexDataContainer and DataContainerTable are helper implementations on top
of the basic CUBA data layout.  As long the as the CUBA keywords spelling
does not change or CUBA keywords are removed we do not need to change the
version. We need however, to make sure that the IndexedDataContainer and
DataContainerTable are flexible enough to handle such behaviour.

>  Should we address that in this issue or open up another issue?
>
We could for example add a version to cuba.yml and the cuds_version for
> IndexedDataContainer and DataContainerTable could follow that number.
>
While at first sight it should not be a problem I think we should be ok.
You make a good point what we need to investigate and write down a plan on
all cases we can think and see if the proposed implementation covers us.

**Comment by itziakos** [comment]
The usecase list will be the guidelines for the developer documentation on what needs to happen when something changes.

**Comment by nathanfranklin** [comment]
> IndexDataContainer` and `DataContainerTable` are helper implementations on top of the basic CUBA data hdf5 **layout*. As long the as the CUBA keywords spelling does not change or CUBA keywords are removed we do not need to change the version. We need however, to make sure that the `IndexedDataContainer` and `DataContainerTable` are flexible enough to handle such behavior.

The changing of the version system would would not be limited to just spelling CUBA keywords or the removal of CUBA keywords.  The addition of CUBA keywords would also require that the version numbers be changed (i.e. incremented)l.  Without having incremented the version in such a case, we would encounter some problems.  For example, we could still read the file in the current implementation of `IndexedDataContainer` and `DataContainerTable` but we couldn't be able to use the newly added CUBA keywords with such a file.  If we augmented `IndexedDataContainer` and `DataContainerTable` so that they recognized this situation and then update the table description to use the larger list of CUBA keywords, then we still have a problem as older software unaware of the changes to CUBA would fail even though the file version said it shouldn't.

> While at first sight it should not be a problem (I think we should be ok). You make a good point that we need to investigate and write down a plan on all cases we can think and see if the proposed implementation covers us.

> The list of the various scenarios will help us write guidelines for the developer documentation on what needs to happen when something changes.

Sounds like a good idea.  

For the deadline of 0.2.0 release of Monday, I will follow your original proposal of having a  `cuds_version` for the root and for the container layouts.  Then either open an issue or wiki page to follow this.

**Comment by itziakos** [comment]
> The addition of CUBA keywords would also require that the version numbers be changed (i.e. incremented).

The intention for the CUDS Containers is to have a predefined set of CUBA keys that they store which is defined at runtime (see proposal https://publicwiki-01.fraunhofer.de/SimPhoNy-Project/index.php/Supported_CUBA_keys_for_container). So in the case were one has an additional CUBA keyword, the old container will be treated like a container with a reduced set of CUBA keywords. The layout will not change and all the information is still available.

**Comment by nathanfranklin** [comment]
Yes, I understand that reading the stored information would still work;  I am just saying that writing to the CUDS file using the newly defined CUBA keyword would fail.

**Comment by itziakos** [comment]
> Yes, I understand that reading the stored information would still work; I am just saying that writing to the CUDS file using the newly defined CUBA keyword would fail.

I would think that using a newly defined CUBA keyword in a DataContainer and trying to save it in a old CUDS file is exactly the same as trying to save a CUBA keywords in an engine that does not support it. The new keyword will be ignored (with a appropriate warning). The underline implementation of the HDF5 code is designed to support a predefined set of CUBA record. Currently it is true that we do not expose this functionality but it is there.

If we would like to support extending the information in a CUDS container backed by an HDF5 file then we can add that functionality (it is simple to add columns to a table). This feature will be the same if the CUBA keyword was a newly defined one or on that was initially not stored. However, extending a table is not necessary changing the layout. The CUBA values are still stored in a table columns named after CUBA.name and there is a mask column that shows missing values.

**Comment by nathanfranklin** [comment]
> However, extending a table is not necessary changing the layout

I understand that the layout doesn't change but  I am still maintaining my earlier comment that changes of CUBA (even if its just the addition of keywords) need to be reflected in our use of `cuds_version`` (by either just always incrementing the 'cuds_version' of the file when keywords are added, renamed, deleted).     

For example, if we don't change the 'cuds_version' when we add new keywords in a release Y which replaces X, there is a problem.   Y can succesfully read the file and can support writing to the file with the new keywords by extending the table.  But now, a user of SimPhoNy version X can no longer read the file even though his software thinks it supports this 'cuds_version'.

**Comment by nathanfranklin** [comment]
> I would think that using a newly defined CUBA keyword in a DataContainer and trying to save it in a old CUDS file is exactly the same as trying to save a CUBA keyword in an engine that does not support it. The new keyword will be ignored (with an appropriate warning). The underline implementation of the HDF5 code is designed to support a predefined set of CUBA records and save only those. Currently it is true that we do not expose this functionality but it is there.

Yeah, this a good approach once when the proposal is incorporated.  For the time-being, we can just ensure that `cuds-version` is incremented when any affected changes to the CUBA keywords occur.

**Comment by itziakos** [comment]
Ok, I see your point, I was working on top of the cuba keys solution.
